### PR TITLE
add log4j2 extension

### PR DESCRIPTION
**Pull request by wanwanpp**
## What is the purpose of the change
#1713 
to add log4j2 for dubbo.

## Brief changelog

add log4j2 framework.

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [x] Run `mvn clean install -DskipTests` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [x] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by noahziheng** [comment]
Log4j2LoggerAdapter has two problems:
 - Collection is useless.
 - `SetLevel` and `SetFile` isn't implement.

I finished review on `file changed`:)

File: dubbo-common/src/main/java/org/apache/dubbo/common/logger/log4j2/Log4j2LoggerAdapter.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    private File file;
+
+    public Log4j2LoggerAdapter() {
+        try {
+            org.apache.logging.log4j.Logger logger = LogManager.getRootLogger();
+            if (logger != null) {
+                //Appender
+                LoggerContext context = (LoggerContext)LogManager.getContext(false);
+                Map<String, Appender> appenders = context.getConfiguration().getAppenders();
+                Collection<Appender> appenderCollection = appenders.values();
```
Comment (noahziheng) [inline dubbo-common/src/main/java/org/apache/dubbo/common/logger/log4j2/Log4j2LoggerAdapter.java]: Collection is useless.

File: dubbo-common/src/main/java/org/apache/dubbo/common/logger/log4j2/Log4j2LoggerAdapter.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        return new Log4j2Logger(LogManager.getLogger(key));
+    }
+
+    @Override
+    public Level getLevel() {
+        return fromLog4jLevel(LogManager.getRootLogger().getLevel());
+    }
+
+    @Override
+    public void setLevel(Level level) {
```
Comment (noahziheng) [inline dubbo-common/src/main/java/org/apache/dubbo/common/logger/log4j2/Log4j2LoggerAdapter.java]: `SetLevel` method isn't implemented,

File: dubbo-common/src/main/java/org/apache/dubbo/common/logger/log4j2/Log4j2LoggerAdapter.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    @Override
+    public File getFile() {
+        return file;
+    }
+
+    @Override
+    public void setFile(File file) {
```
Comment (noahziheng) [inline dubbo-common/src/main/java/org/apache/dubbo/common/logger/log4j2/Log4j2LoggerAdapter.java]: `SetLevel` method isn't implemented,

File: dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java [hunk 4]
```diff
@@ -47,6 +48,8 @@
             setLoggerAdapter(new Log4jLoggerAdapter());
         } else if ("jdk".equals(logger)) {
             setLoggerAdapter(new JdkLoggerAdapter());
+        } else if ("log4j2".equals(logger)) {
+            setLoggerAdapter(new Log4j2LoggerAdapter());
```
Comment (diecui1202) [inline dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java]: IMO, log4j2 could place before jdk log

File: dubbo-common/src/main/java/org/apache/dubbo/common/logger/log4j2/Log4j2LoggerAdapter.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+public class Log4j2LoggerAdapter implements LoggerAdapter {
+
+    private File file;
+
+    public Log4j2LoggerAdapter() {
+        try {
+            org.apache.logging.log4j.Logger logger = LogManager.getRootLogger();
+            if (logger != null) {
+                LoggerContext context = (LoggerContext)LogManager.getContext(false);
+                Map<String, Appender> appenders = context.getConfiguration().getAppenders();
```
Comment (diecui1202) [inline dubbo-common/src/main/java/org/apache/dubbo/common/logger/log4j2/Log4j2LoggerAdapter.java]: Pls check NPE for appenders.

File: dubbo-common/pom.xml [hunk 6]
```diff
@@ -43,6 +43,14 @@
...
             <artifactId>log4j</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-core</artifactId>
+        </dependency>
```
Comment (diecui1202) [inline dubbo-common/pom.xml]: add provided scope for these 2 dependencies?

**Comment by diecui1202** [comment]
@wanwanpp please check the comments and review suggestions.

**Comment by diecui1202** [comment]
The two methods getLevel and getFile are used for telnet handler, but I found it's not correct. 

Therefore in this PR, the two methods are just empty implementation, I've started a new issue #2312 to track this.

**Pull request merged**